lsi/rom/tools/cp16mic.py

```diff
@@ -0,0 +1,981 @@
+                  (self.lnum, self.lc, val, line), file=flst)
+            for emsg in self.elst:
+                print("%s" % emsg, file=flst)
+        return
+
+    #
+    # Set the pass number and reinit the parser
+    #
+    def set_pass(self, ps):
+        self.ps = ps
+        self.lc = 0
+        self.ecnt = 0
+        self.wcnt = 0
+        self.lnum = 0
+        self.clin = 0
+        self.cloc = self.bloc.get(0, {})
+        return
+
+    #
+    # Process one source file
+    #
+    def do_assembly(self, fsrc, flst):
+        self.fend = 0
+        for line in fsrc:
+            self.lnum += 1
+            line = line.strip('\r\n')
+
+            try:
+                self.do_line(line)
+            except SyntaxError as err:
+                self.log_error(err)
+
+            if self.ps == 2 and flst is not None:
+                self.do_list(line, flst)
+            if self.fend:
+                break
+        self.local_block()
+        if self.fend == 0:
+            self.elst = []
+            self.log_warning("missing '.end' directive")
+            if flst is not None:
+                print('%s' % self.elst[0], file=flst)
+        return
+
+    def write_obj(self, fobj, width):
+        msk = (1 << width) - 1
+        lim = -1
+        for i in range(MAX_LC):
+            if self.data[i] >= 0:
+                lim = i
+        if lim < 0:
+            return
+        lim += 1
+        print('DEPTH = %d;\r\n'
+              'WIDTH = %d;\r\n'
+              'ADDRESS_RADIX = HEX;'
+              'DATA_RADIX = HEX;'
+              'CONTENT BEGIN' % (lim, width), file=fobj)
+        for i in range(lim):
+            data = self.data[i]
+            if data >= 0:
+               print('%03X: %06X;' % (i, data & msk), file=fobj)
+        print('END;',  file=fobj)
+        return
+
+    def write_ttl(self, fttl):
+        msk = 0x0f << 18
+        for i in range(MAX_LC):
+            data = self.data[i]
+            if data >= 0 and data & msk:
+                data &= msk
+                print('%03X: %06X, %1X;' % (i, data, data >> 18), file=fobj)
+        return
+
+def createParser ():
+    p = argparse.ArgumentParser(
+        description='Western Digital MCP-1600 Microcode Assembler, '
+                    'Version 20.05a, (c) 1801BM1')
+    p.add_argument ('src', nargs='+',
+                    help='input source file(s)', metavar='file [file ...]')
+    p.add_argument ('-l', '--lst', help='output listing file', metavar='file')
+    p.add_argument ('-o', '--obj', help='output object file', metavar='file')
+    p.add_argument ('-t', '--ttl', help='output TTL-logic file', metavar='file')
+    return p
+ 
+if __name__ == '__main__':
+    parser = createParser()
+    params = parser.parse_args()
+
+    try:
+        asm = wd16()
+        #
+        # Check all source files existance
+        # 
+        for src in params.src:
+            fsrc = asm.open_file(src, 'mic', 'r')
+            asm.close_file(fsrc)
```

Comment:
in POSIX you have special API `access()` to check that file exists w/o opening it. In Python you have `os.path.exists(filename)` function

Reply:
The objective is to check everything, including the access rights